content/blog/2024/2024-05-25-creating-an-npm-package .md

```diff
@@ -0,0 +1,534 @@
+
+Mastering the art of creating NPM packages is crucial. It allows for the encapsulation of reusable code, simplifies project development, and promotes collaboration by sharing useful libraries with the community. This accelerates the development process and gives us the option to keep this package private or make it public (open-source) for others to use.
+
+# Prerequisites
+We'll need the following:
+- [Node.js](https://nodejs.org/) installed on our computer.
+- Basic knowledge of TypeScript.
+- [GitHub Account](https://github.com/).
+- [NPM Account](https://www.npmjs.com/signup).
+
+In this post, we will create an NPM package to validate user inputs such as emails, mobile numbers, and social media links. Instead of rewriting these validation functions for each project, our package can be installed across projects to simplify and standardize the validation process.
+
+Let's jump right into it:
+
+{{% github "https://github.com/thombergs/code-examples/tree/master/nodejs/node-create-npm-pacakage" %}}
+
+# Step 1: Setting up Node.js Application
+Begin by creating an empty folder named `validate-npm-pc`.
+
+This folder will serve as the name of our NPM package. It's important to choose a unique name since the NPM registry hosts a vast number of packages, each requiring a distinct name for successful publication.
+
+To initialize Node.js in the project, run the following command in the terminal:
+
+```bash
+npm init -y
+```
+This command generates a `package.json` file, which holds essential metadata about our project, including details of the dependencies and version information.
+
+Again in the terminal, we will execute the following:
+```bash
+mkdir -p .github/workflows src __tests__ 
+touch src/index.ts src/validate.ts __tests__/validate.ts .github/workflows/release.yml .gitignore
+```
+These commands create all the necessary files and folders required in the package.
+
+Install all the dependencies needed in the package by running the following command:
+```bash
+npm install typescript jest ts-jest @types/jest @changesets/cli  --save-dev
+```
+Here's a brief overview of what each dependency does:
+- `typescript`: Enables static typing in our code.
+- `jest`: A JavaScript testing framework.
+- `ts-jest`: A TypeScript preprocessor with source map support for Jest.
+- `@types/jest`: Provides TypeScript type definitions for Jest.
+- `@changesets/cli`: is a command-line tool for managing versioning and changelogs in a monorepo setup. Changeset automates the NPM versioning process for our package.
+
+NPM versioning follows the [Semantic Versioning](https://semver.org/) (SemVer) convention, which consists of three numbers separated by periods: `MAJOR.MINOR.PATCH` [1.2.3].
+
+![npm-versioning](https://i.imgur.com/TUziofC.png)
+{{% image alt="npm-versioning" src="images/posts/node-create-npm-package/npm-versioning.png" %}}
```

Comment:
Where is this image from, given that there is an imgur link? I removed the duplicate, by the way.

Reply:
Also, this image is a bit hard to understand. It would make much more sense if the the upper bubble contained version 1.0.0, for example, and the others would contain 1.0.1 for patch, 1.1.0 for minor, and 2.0.0 for major.

Reply:
Yes, i created the image myself.
Your versioning example seems better, I have updated the image. Thank you